STModCellular.cpp

```diff
@@ -18,53 +18,104 @@
 #include "STModCellular.h"
 #include "mbed_wait_api.h"
 #include "mbed_trace.h"
+#include "DigitalIn.h"
 
 #define TRACE_GROUP "CELL"
 
 using namespace mbed;
 
-STModCellular::STModCellular(FileHandle *fh) : QUECTEL_UG96(fh),
+STModCellular::STModCellular(FileHandle *fh) : STMOD_CELLULAR_MODEM(fh),
     m_powerkey(MBED_CONF_STMOD_CELLULAR_POWER),
     m_reset(MBED_CONF_STMOD_CELLULAR_RESET),
     m_simsel0(MBED_CONF_STMOD_CELLULAR_SIMSEL0),
     m_simsel1(MBED_CONF_STMOD_CELLULAR_SIMSEL1),
     m_mdmdtr(MBED_CONF_STMOD_CELLULAR_MDMDTR)
 {
-    // start with modem disabled
-    m_mdmdtr.write(0);
-    m_simsel0.write(MBED_CONF_STMOD_CELLULAR_SIM_SELECTION);
-    m_simsel1.write(0);
 
+    tr_debug("STModCellular creation\r\n");
+
+    /* Ensure PIN SIMs are set as input */
+    DigitalIn sim_reset(MBED_CONF_STMOD_CELLULAR_SIM_RESET);
+    DigitalIn sim_clk(MBED_CONF_STMOD_CELLULAR_SIM_CLK);
+    DigitalIn sim_data(MBED_CONF_STMOD_CELLULAR_SIM_DATA);
+
+    // start with modem disabled
+    m_powerkey.write(0);
     m_reset.write(1);
-    wait_ms(250);
+    wait_ms(200);
     m_reset.write(0);
+    wait_ms(150);
+
+    wait_ms(50);
+    m_simsel0.write(MBED_CONF_STMOD_CELLULAR_SIM_SELECTION & 0x01);
+    m_simsel1.write(MBED_CONF_STMOD_CELLULAR_SIM_SELECTION & 0x02);
+    wait_ms(50);
 }
 
 STModCellular::~STModCellular()
 {
 }
 
 nsapi_error_t STModCellular::soft_power_on() {
-    nsapi_error_t err = QUECTEL_UG96::soft_power_on();
+    tr_debug("STMOD cellular modem power ON\r\n");
+
+#if (MBED_CONF_STMOD_CELLULAR_TYPE == STMOD_UG96)
+    tr_debug("Booting UG96\r\n");
+    m_reset.write(1);
+    wait_ms(200);
+    m_reset.write(0);
+    wait_ms(150);
+    m_powerkey.write(1);
+    wait_ms(150);
+    m_powerkey.write(0);
+    /* Because modem status is not available on STMOD+ connector,
+     * let's wait for Modem complete boot */
+    wait_ms(2300);
```

Comment:
Is this `wait_ms` required since we are waiting for the RDY signal ?

Reply:
I think it is required - I understand that reading UART before this timing may be unpredictable.  This is the reliable sequence that I was given to and should be followed (As stated in the comment, if modem_status pin would be available, we would use this signal before reading uart, but it isn't)